retworkx/__init__.py

```diff
@@ -769,7 +769,34 @@ def _graph_k_shortest_path_lengths(graph, start, k, edge_cost, goal=None):
 
 @functools.singledispatch
 def dfs_edges(graph, source=None):
-    """Get edge list in depth first order
+    """Get an edge list of the tree edges from a depth-first traversal
+
+    The pseudo-code for the DFS algorithm is listed below. The output
+    contains the tree edges found by the procedure.
+
+    ::
+
+        DFS(G, s)
+          for each vertex u in V
+              color[u] := WHITE
+          end for
+          STACK := Ø
+          color[s] := GRAY
+          PUSH(STACK, s)                           start at vertex s
+          while (STACK != Ø)
+              u := POP(STACK)
+              for each vertex v in Adj[u]          (u,v) is a tree edge
+                  if (color[v] = WHITE)
+                      color[v] = GRAY
+                      PUSH(STACK, v)
+              end for
+              color[u] := BLACK                    finish vertex u
+          end while
+
```

Comment:
Actually the pseudocode does not seem to be fully accurate, i.e it doesn't perform a standard DFS.

We can use the pseudocode of a recursive implementation like here https://github.com/Qiskit/retworkx/blob/d84d10372927b9a116cea0e1b73796f315d8270a/retworkx/__init__.py#L1746-L1766

or an iterative one which closely resembles the implementation of this function:

```
DFS(G, v) is
    let S be a stack
    PUSH(S, (v, iterator of G.edges(v)))
    while (S != Ø)
        let (v, iterator) := LAST(S)
        if hasNext(iterator) then
            w := next(iterator)
            if w is not labeled as discovered then
                label w as discovered                       # (u, w) is a tree edge
                PUSH(S, (w, iterator of G.edges(w)))
        else
            POP(S)
```

Reply:
It's not the standard but if you look at the function, it does a stack based DFS

Reply:
But if I run the pseudocode with a triangle as a graph starting from node 0, it outputs (0, 1), (0, 2) as tree edges but the DFS output is (0, 1), (1, 2).

Reply:
Indeed for some reason that function has a weird DFS, but it is a DFS. I updated the docstring to use the pseudocode you provided.